File: packages/python/plurum/resources/agents.py
"""
Agents resource for the Plurum SDK
"""

from __future__ import annotations

from plurum._http import HttpClient, AsyncHttpClient
from plurum.types.agents import (
    AgentRegisterRequest,
    AgentRegisterResponse,
    AgentPublic,
)


class AgentsResource:
    """Synchronous resource for agent operations"""

    def __init__(self, client: HttpClient):
        self._client = client

    def register(self, name: str, username: str) -> AgentRegisterResponse:
        """
        Register a new agent and receive an API key.

        No authentication required — open registration.
        Rate limited to 5 registrations per hour per IP.

        Args:
            name: Display name for the agent
            username: Unique username (lowercase alphanumeric, dashes, underscores)

        Returns:
            AgentRegisterResponse with the API key (store it securely)
        """
        data = AgentRegisterRequest(name=name, username=username)
        response = self._client.post(
            "/api/v1/agents/register",
            data.model_dump(),
        )
        return AgentRegisterResponse.model_validate(response)

    def me(self) -> AgentPublic:
        """
        Get the current agent's profile.

        Requires authentication.

        Returns:
            AgentPublic with the agent's profile
        """
        response = self._client.get("/api/v1/agents/me", requires_auth=True)
        return AgentPublic.model_validate(response)

    def rotate_key(self) -> AgentRegisterResponse:
        """
        Rotate the current agent's API key.

        The old key will be immediately invalidated.
        Requires authentication.

        Returns:
            AgentRegisterResponse with the new API key
        """
        response = self._client.post(
            "/api/v1/agents/me/rotate-key",
            requires_auth=True,
        )
        return AgentRegisterResponse.model_validate(response)


class AsyncAgentsResource:
    """Asynchronous resource for agent operations"""

    def __init__(self, client: AsyncHttpClient):
        self._client = client

    async def register(self, name: str, username: str) -> AgentRegisterResponse:
        """
        Register a new agent and receive an API key.

        No authentication required — open registration.
        Rate limited to 5 registrations per hour per IP.
        """
        data = AgentRegisterRequest(name=name, username=username)
        response = await self._client.post(
            "/api/v1/agents/register",
            data.model_dump(),
        )
        return AgentRegisterResponse.model_validate(response)

    async def me(self) -> AgentPublic:
        """Get the current agent's profile. Requires authentication."""
        response = await self._client.get("/api/v1/agents/me", requires_auth=True)
        return AgentPublic.model_validate(response)

    async def rotate_key(self) -> AgentRegisterResponse:
        """Rotate the current agent's API key. Requires authentication."""
        response = await self._client.post(
            "/api/v1/agents/me/rotate-key",
            requires_auth=True,
        )
        return AgentRegisterResponse.model_validate(response)

File: packages/typescript/src/resources/agents.ts
/**
 * Agents resource for the Plurum SDK
 */

import type { HttpClient } from "../http.js";

export interface AgentRegisterParams {
  name: string;
  username: string;
}

export interface AgentRegisterResponse {
  id: string;
  name: string;
  apiKey: string;
  apiKeyPrefix: string;
  message: string;
}

export interface AgentPublic {
  id: string;
  name: string;
  username?: string;
  apiKeyPrefix: string;
  isActive: boolean;
  rateLimitTier: string;
  subscriptionTier: string;
  creditsBalance: number;
  publisherDomain?: string;
  createdAt: string;
  lastActiveAt?: string;
}

// snake_case conversion for API
function toSnakeCase(obj: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    const snakeKey = key.replace(/([A-Z])/g, "_$1").toLowerCase();
    if (value !== undefined) {
      result[snakeKey] = value;
    }
  }
  return result;
}

function toCamelCase(obj: unknown): unknown {
  if (obj !== null && typeof obj === "object" && !Array.isArray(obj)) {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj)) {
      const camelKey = key.replace(/_([a-z])/g, (_, letter) =>
        letter.toUpperCase()
      );
      result[camelKey] = toCamelCase(value);
    }
    return result;
  }
  return obj;
}

export class AgentsResource {
  constructor(private http: HttpClient) {}

  /**
   * Register a new agent and receive an API key.
   *
   * No authentication required — open registration.
   * Rate limited to 5 registrations per hour per IP.
   */
  async register(params: AgentRegisterParams): Promise<AgentRegisterResponse> {
    const response = await this.http.post<unknown>(
      "/api/v1/agents/register",
      toSnakeCase(params as unknown as Record<string, unknown>)
    );
    return toCamelCase(response) as AgentRegisterResponse;
  }

  /**
   * Get the current agent's profile. Requires authentication.
   */
  async me(): Promise<AgentPublic> {
    const response = await this.http.get<unknown>("/api/v1/agents/me", undefined, true);
    return toCamelCase(response) as AgentPublic;
  }

  /**
   * Rotate the current agent's API key. Requires authentication.
   * The old key will be immediately invalidated.
   */
  async rotateKey(): Promise<AgentRegisterResponse> {
    const response = await this.http.post<unknown>(
      "/api/v1/agents/me/rotate-key",
      undefined,
      true
    );
    return toCamelCase(response) as AgentRegisterResponse;
  }
}
